### DISCUSSION TITLE
Define labels from REST API data

### DISCUSSION THREAD

**Discussion by JoanePouech** [grafana/grafana-infinity-datasource#728] [Q&A] [answered]
Hi,
First, thank you for this very useful and powerful plugin !
I have a question about how to format data to differientiate fields and labels.
For example, I receive this kind of times series data from my API : 

``` 
[
  {
    timestamp: '2023-09-22T09:30:00.000000Z',
    temperature: 15.8,
    humidity: 45,
    town: 'Paris'
  },
  {
    timestamp: '2023-09-22T09:40:00.000000Z',
    temperature: 16.1,
    humidity: 52,
    town: 'Paris'
  },
  {
    timestamp: '2023-09-22T09:30:00.000000Z',
    temperature: 12.4,
    humidity: 73,
    town: 'London'
  },
  {
    timestamp: '2023-09-22T09:40:00.000000Z',
    temperature: 13.4,
    humidity: 70,
    town: 'London'
  },
  // etc
]
```
In this example, `temperature` and `humidity` are fields, and `town` should be interpreted as label but Grafana doesn't.
I will have on my panel a unique curve for temperature or humidity, whatever the town. And I don't have access to the town information in this curve, as it's interpreted by Grafana as a separate time series.
I searched in the options of the plugin, I have tried several return data format for the API, but I didn't find how to resolve this problem.

Is there a way in Infinity plugin or a data format (I have the possibility to modify the return data of the API) to make Grafana understanding which key is a field or a label in my data ?

**Comment by yesoreyeram** [comment]
In query editor, have you tried setting the format to timeseries? That should do the trick.

**Reply by JoanePouech** [reply]
Thank you for answer !
Yes, I tried, but I have a 'No data' result in Table view panel 
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/36c8d907-8747-4565-90d7-49abff6632f8)

With Data Frame or Table format, the columns are well defined, but I would like to specify the "field-columns" and the "labels-columns"
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/bc8a91e8-ade6-4cf9-aa5c-5a1689388dd8)

With the above example, I obtain 2 curves (for temperature and humidity), by I would like 4 (temperature-Paris, temperature-London, humidity-Paris, humidity-London)
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/effe330d-ea36-4bf9-aa53-b5f07b8e0930)

**Comment by JoanePouech** [comment, marked as answer]
I found a way to obtain what I need, I share my solution here as it could be useful to other people.

I formatted my data like that:
```
{
  fields: [
    { name: 'timestamp', type: 'time', values: ['2023-09-22T09:30:00.000000Z', '2023-09-22T09:40:00.000000Z'] },
    {
      name: 'temperature',
      values: [15.8, 16.1],
      labels: {
        town: 'Paris',
      },
    },
    {
      name: 'temperature',
      values: [12.4, 13.4],
      labels: {
        town: 'London',
      },
    },
    {
      name: 'humidity',
      values: [45, 52],
      labels: {
        town: 'Paris',
      },
    },
    {
      name: 'humidity',
      values: [73, 70],
      labels: {
        town: 'London',
      },
    },
  ],
};
```

and I set the format to "As Is" in the query editor.

I obtain 4 series and 4 curves in Grafana :
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/4d05bae1-de87-4683-a134-d3a98d35d51d)
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/5477442e-b537-4814-bc2a-578be838f082)

and I can define dynamically labels for captions : 
![image](https://github.com/grafana/grafana-infinity-datasource/assets/70484783/3badd922-28c6-4c82-8739-9bf5c28c21ad)

**Reply by oskarm93** [reply]
@JoanePouech Would you be able to share more on this set up? I copy pasted your sample data, set Parser to Backend>JSONata, Format: As-Is and no other parsing options etc and I was not able to get the time series graph to show anything.
